src/api.rs

```diff
@@ -0,0 +1,243 @@
+use crate::{
+    auth::{Auth, AuthQuery, IDToken, Service},
+    channel::Channel,
+    hub::{Hub, HubMember},
+    is_valid_username, new_id,
+    permission::HubPermission,
+    user::{GenericUser, User},
+    Error, Result, AUTH, ID,
+};
+
+pub async fn start_login(service: Service) -> String {
+    Auth::start_login(AUTH.clone(), service).await
+}
+
+pub async fn complete_login(service: Service, query: AuthQuery) -> Result<IDToken> {
+    Auth::handle_oauth(AUTH.clone(), service, query).await
+}
+
+pub async fn invalidate_tokens(user: &User) {
+    Auth::invalidate_tokens(AUTH.clone(), user.id).await
+}
+
+pub async fn get_user_stripped(id: ID) -> Result<GenericUser> {
+    if let Ok(other) = User::load(&id).await {
+        Ok(other.to_generic())
+    } else {
+        Err(Error::UserNotFound)
+    }
+}
+
+pub async fn change_username(user: &mut User, new_name: String) -> Result<String> {
+    is_valid_username(&new_name)?;
+    let old_name = user.username.clone();
+    user.username = new_name;
+    user.save().await?;
+    Ok(old_name)
+}
+
+pub async fn create_hub(owner: &mut User, name: String) -> Result<ID> {
+    is_valid_username(&name)?;
+    let mut id = new_id();
+    while Hub::load(&id).await.is_ok() {
+        id = new_id();
+    }
+    let mut new_hub = Hub::new(name, id.clone(), &owner);
+    let channel_id = new_hub.new_channel(&owner.id, "chat".to_string()).await?;
+    if let Some(group) = new_hub.groups.get_mut(&new_hub.default_group) {
+        group.set_channel_permission(
+            channel_id.clone(),
+            crate::permission::ChannelPermission::ViewChannel,
+            crate::permission::PermissionSetting::TRUE,
+        );
+        group.set_channel_permission(
+            channel_id.clone(),
+            crate::permission::ChannelPermission::SendMessage,
+            crate::permission::PermissionSetting::TRUE,
+        );
+        group.set_channel_permission(
+            channel_id.clone(),
+            crate::permission::ChannelPermission::ReadMessage,
+            crate::permission::PermissionSetting::TRUE,
+        );
+    }
+    owner.in_hubs.push(id.clone());
+    owner.save().await?;
+    new_hub.save().await?;
+    Ok(id)
+}
+
+pub async fn get_hub(user: &User, hub_id: &ID) -> Result<Hub> {
+    user.in_hub(&hub_id)?;
+    let hub = Hub::load(hub_id).await?;
+    hub.strip(&user.id)
+}
+
+pub async fn delete_hub(user: &User, hub_id: &ID) -> Result<()> {
+    user.in_hub(&hub_id)?;
+    let hub = Hub::load(hub_id).await?;
+    let member = hub.get_member(&user.id)?;
+    if member.has_all_permissions() {
+        if let Ok(()) = tokio::fs::remove_file(hub.get_info_path()).await {
+            if let Ok(()) = tokio::fs::remove_dir_all(hub.get_data_path()).await {
+                Ok(())
+            } else {
+                Err(Error::DeleteFailed)
+            }
+        } else {
+            Err(Error::DeleteFailed)
+        }
+    } else {
+        Err(Error::NoPermission)
+    }
```

Comment:
You might want to include the sources of your errors to help you debug. It might be a pain to maintain though. If you have a boxed dynamic Error, you'd need it to be `dyn Error + Send + Sync` I think, or even `'static`.

Also, if you think writing one-liners can be kind of messy, you can dismiss some of my suggestions.

Reply:
What do you mean by including the source? As in, including the permission that is missing?

Reply:
The source of your error can be from an external error e.g. a tokio error. You might want to include that in a box but it can be hard to do. Just something to think about.

Reply:
Ah ok, yeah, I probably won't do that...